### DISCUSSION TITLE
Dudas con la governanza

### DISCUSSION THREAD

**Discussion by franciscomirasg** [opencanarias/taple-client#50] [Q&A] [answered]
Estimado equipo de TAPLE,

Es un placer dirigirme a ustedes. Actualmente, estoy llevando a cabo un Trabajo de Fin de Grado en el cual utilizaré TAPLE como parte fundamental de mi caso de estudio. Durante mi investigación, he revisado detalladamente la documentación disponible en su página web sobre cómo escribir las gobernanzas, y me han surgido algunas interrogantes que me gustaría aclarar:

1. Respecto a la definición de los miembros, me gustaría saber si se permite utilizar el campo "definitions" de JSON SCHEMA para almacenar información adicional, como subesquemas y enumerados.

2. Si el quorum se establece en 0.0, ¿cuál sería el efecto resultante? ¿Es posible utilizar esta configuración para eliminar la necesidad de aprobadores o validadores?

3. Me gustaría obtener una mayor claridad en cuanto a la declaración de los testigos. Mi objetivo es asegurar que cada ciudadano pueda acceder únicamente a su propia información y cartera de puntos, sin tener acceso a los datos de otros ciudadanos. Por otro lado, el ayuntamiento debería tener acceso a la información de todos los ciudadanos, mientras que la empresa de reciclaje no debería tener acceso a la información personal de los ciudadanos.

Agradezco de antemano su atención y orientación en relación a estas cuestiones.

Adjunto la gobernanza por si ayuda:
```json
{
	"members": [
		{
			"id": "b11a2304-c1a0-42b4-bafe-f22c73812051",
			"tags": {},
			"description": "Citizen 1",
			"key": "REDACTED"
		},
		{
			"id": "49038391-f970-4fae-90c7-dece4db3c131",
			"tags": {},
			"description": "Citizen 2",
			"key": "REDACTED"
		},
		{
			"id": "City Hall",
			"tags": {},
			"description": "City Hall",
			"key": "REDACTED"
		},
		{
			"id": "32c069c5-6ed7-4193-988c-0fc1e029552c",
			"tags": {},
			"description": "Container 1",
			"key": "REDACTED"
		},
		{
			"id": "abf385e8-fbdd-433d-80f6-1e6e16e42c87",
			"tags": {},
			"description": "Container 2",
			"key": "REDACTED"
		},
		{
			"id": "Trash Company",
			"tags": {},
			"description": "Trash Company",
			"key": "REDACTED"
		}
	],
	"schemas": [
		{
			"id": "Citizen",
			"tags": {},
			"content": {
				"type": "object",
				"properties": {
					"identidad": {
						"title": "Identidad",
						"description": "Identidad del ciudadano",
						"allOf": [
							{
								"$ref": "#/definitions/Identidad"
							}
						]
					},
					"direccion": {
						"title": "Direccion",
						"description": "Direcion del ciudadano",
						"allOf": [
							{
								"$ref": "#/definitions/Direcion"
							}
						]
					},
					"cuid": {
						"title": "Cuid",
						"description": "Codigo de identificaion unico del ciudadano",
						"maxLength": 36,
						"minLength": 36,
						"type": "string"
					}
				},
				"required": [
					"identidad",
					"direccion"
				],
				"definitions": {
					"Identidad": {
						"title": "Identidad",
						"description": "Data model for the identity of a citizen\n\natributes:\n\n    - nombre: str. Name\n    - apellidos: str. Surnames\n    - ni: str. DNI",
						"type": "object",
						"properties": {
							"nombre": {
								"title": "Nombre",
								"description": "Nombre del ciudadano",
								"type": "string"
							},
							"apellidos": {
								"title": "Apellidos",
								"description": "Primer apellido del ciudadano",
								"type": "string"
							},
							"ni": {
								"title": "Ni",
								"description": "DNI del ciudadano",
								"maxLength": 9,
								"minLength": 9,
								"type": "string"
							}
						},
						"required": [
							"nombre",
							"apellidos",
							"ni"
						]
					},
					"ViaType": {
						"title": "ViaType",
						"description": "Enumerate for the type of street",
						"enum": [
							"Calle",
							"Avenida",
							"Plaza",
							"Paseo",
							"Carretera",
							"Camino",
							"Alameda",
							"Otro"
						],
						"type": "string"
					},
					"Direcion": {
						"title": "Direcion",
						"description": "Data model for the address of a citizen\n\natributes:\n    - type: ViaType. Type of street\n    - direccion: str. Address\n    - codigo_postal: str. Postal code\n    - comunidad: str. Community\n    - ciudad: str. City",
						"type": "object",
						"properties": {
							"type": {
								"description": "Tipo de v\u00eda",
								"allOf": [
									{
										"$ref": "#/definitions/ViaType"
									}
								]
							},
							"direccion": {
								"title": "Direccion",
								"description": "Direccion",
								"type": "string"
							},
							"codigo_postal": {
								"title": "Codigo Postal",
								"description": "Codigo Postal",
								"maxLength": 5,
								"minLength": 5,
								"type": "string"
							},
							"comunidad": {
								"title": "Comunidad",
								"description": "Comunidad/Region/Provincia",
								"type": "string"
							},
							"ciudad": {
								"title": "Ciudad",
								"description": "Ciudad",
								"type": "string"
							}
						},
						"required": [
							"type",
							"direccion",
							"codigo_postal",
							"comunidad",
							"ciudad"
						]
					}
				}
			}
		},
		{
			"id": "RecycleOperation",
			"tags": {},
			"content": {
				"type": "object",
				"properties": {
					"type": {
						"description": "Tipo de residuo",
						"allOf": [
							{
								"$ref": "#/definitions/TrashType"
							}
						]
					},
					"peso": {
						"title": "Peso",
						"description": "Peso de los residuos, en gramos",
						"exclusiveMinimum": 0,
						"type": "number"
					},
					"container": {
						"title": "Container",
						"description": "Id del contenedor",
						"maxLength": 36,
						"minLength": 36,
						"type": "string"
					}
				},
				"required": [
					"type",
					"peso",
					"container"
				],
				"definitions": {
					"TrashType": {
						"title": "TrashType",
						"description": "Enumerate for the type of trash",
						"enum": [
							"Org\u00e1nico",
							"Papel y cart\u00f3n",
							"Vidrio",
							"Pl\u00e1stico",
							"Ropa y textil",
							"Electr\u00f3nicos",
							"Pilas y bater\u00edas",
							"Aceite",
							"No reciclable"
						],
						"type": "string"
					}
				}
			}
		},
		{
			"id": "PointTransaction",
			"tags": {},
			"content": {
				"type": "object",
				"properties": {
					"cuid": {
						"title": "Cuid",
						"description": "Codigo de identificaion unico del ciudadano",
						"maxLength": 36,
						"minLength": 36,
						"type": "string"
					},
					"value": {
						"title": "Value",
						"description": "Valor de la operaci\u00f3n",
						"ne": 0,
						"type": "integer"
					},
					"motivo": {
						"title": "Motivo",
						"description": "Motivo de la operaci\u00f3n",
						"type": "string"
					},
					"balance": {
						"title": "Balance",
						"description": "Balance actual de puntos",
						"default": 0,
						"type": "integer"
					}
				},
				"required": [
					"cuid",
					"value",
					"motivo"
				]
			}
		}
	],
	"policies": [
		{
			"id": "Citizen",
			"validation": {
				"quorum": 1.0,
				"validators": [
					"2daf002ddcf793fe73aa987fbf4aeddfca94bc89164d7a054c9bdb0b25407250"
				]
			},
			"approval": {
				"quorum": 1.0,
				"approvers": [
					"2daf002ddcf793fe73aa987fbf4aeddfca94bc89164d7a054c9bdb0b25407250"
				]
			},
			"invokation": {
				"owner": {
					"allowance": true,
					"approvalRequired": true
				},
				"set": {
					"allowance": false,
					"approvalRequired": true,
					"invokers": []
				},
				"all": {
					"allowance": false,
					"approvalRequired": true
				},
				"external": {
					"allowance": false,
					"approvalRequired": true
				}
			}
		},
		{
			"id": "RecycleOperation",
			"validation": {
				"quorum": 0.0,
				"validators": []
			},
			"approval": {
				"quorum": 0.0,
				"approvers": []
			},
			"invokation": {
				"owner": {
					"allowance": true,
					"approvalRequired": false
				},
				"set": {
					"allowance": true,
					"approvalRequired": false,
					"invokers": [
						"4c8fabf15e73471629997057172e9a375d9073ae9129129128331536950bf58d"
					]
				},
				"all": {
					"allowance": false,
					"approvalRequired": true
				},
				"external": {
					"allowance": false,
					"approvalRequired": true
				}
			}
		},
		{
			"id": "PointTransaction",
			"validation": {
				"quorum": 0.0,
				"validators": []
			},
			"approval": {
				"quorum": 0.0,
				"approvers": []
			},
			"invokation": {
				"owner": {
					"allowance": true,
					"approvalRequired": false
				},
				"set": {
					"allowance": true,
					"approvalRequired": true,
					"invokers": [
						"71a3844e1658cc630cc89b74fc3e9e730be8d7b352c4b8f88014ca9ba742e040",
						"15fb91cc18803632315f1c162f10e9982ab129a7cf91e754445e84f4ff36063e"
					]
				},
				"all": {
					"allowance": false,
					"approvalRequired": true
				},
				"external": {
					"allowance": false,
					"approvalRequired": true
				}
			}
		}
	]
}
```

**Comment by FlorentinPG** [comment, marked as answer]
Saludos Francisco. Te agradecemos tu interés por nuestra tecnología y te deseamos suerte con tu TFG. 

Primeramente comentarte que hemos convertido tu "issue" en esta discusión, pues tu consulta no se correspondía con problemas o fallos detectados en el software. En caso de tener más consultas en el futuro, puedes publicarlas en esta misma discusión o generar una nueva.

Con respecto a las preguntas en sí, quiero empezar anunciándote que dentro de poco se va a producir el lanzamiento de la versión 0.2 de TAPLE, que supone un gran cambio con respecto a la versión 0.1 y sus derivadas. Entre estos cambios, se incluye fundamentalmente la aparición de nuevos actores en la red (evaluadores, aprobadores, validadores y testigos) y contratos inteligentes. Como puedes suponer, estas adiciones supondrán además un cambio en la definición de la gobernanza que ahora es más rica que antes. Teniendo esto en cuenta, es posible que te interese priorizar otros aspectos de tu trabajo hasta que se produzca el lanzamiento oficial. Aún así procederé a responder tus preguntas, en algunos casos haciendo referencia a esta futura versión.

1. No tengo del todo claro a que te refieres en este punto, principalmente porque haces alusión al campo de los miembros. Los miembros tienen un formato específico y este no puede ser alterado. En cuanto a la posibilidad de emplear el campo de "definiciones" como tal, en principio no deberías de tener problema siempre y cuando la entidad siga siendo un JSON SCHEMA válido según el Draft 7. Aún así, su utilidad se vería reducido a estrictamente el schema de los sujetos que estés definiendo. La interoperabilidad entre schemas no es posible. 

2. A partir de la versión 0.2 un Quorum de 0 no será posible, solicitándose mínimo la firma de un actos de la red. Así mismo, en la 0.2 se establecen más alternativas para definir el Quorum deseado. Si quieres eliminar la necesidad de aprobadores, esto es posible hacerlo en la 0.1 a través de la correcta configuración de las policies asociadas a un schema. En la 0.2, por su parte, todos los eventos pasan por un contrato inteligente definido por el usuario, siendo la lógica establecido en el mismo el que decide, bajo el criterio que se prefiera, si la aprobación es o no necesaria, conviertiéndola así en un fase no solo opcional sino también condicional. Por su parte, la validación no debería ser nunca saltada. Si esta se omite, no hay garantía alguna sobre el evento y su credibilidad pasa a verse comprometida. Así mismo, en la 0.1 si quieres compartir un sujeto con alguien, este debe ser validador (en la 0.2 no).

3. En la 0.1 testigo y validador es lo mismo. Además, por las restricciones de la gobernanza, un validador valida todos los sujetos de una gobernanza. Esto complica garantizar cierta privacidad a los datos. En la 0.2 ambos roles se han separado y, además, se ha añadido un sistema de definición de "namespaces" en la gobernanza para poder filtrar los sujetos que realmente son de interés para una entidad. De esta manera no todos los miembros de una misma gobernanza comparten todo. Ese es en principio el nivel de privacidad que vamos a aportar en estos momentos a TAPLE. Si te interesa privacidad dentro de un mismo sujeto, es decir, que algunas atributos del mismo solo pueden ser "vistos" por ciertos miembros de la red, eso en estos momentos no es posible, aunque se trata de una característica que estamos considerando para versiones futuras. Atendiendo a la descripción de tu caso de uso que has aportado creo que con los "namespaces" puedes resolverlo.

Un saludo y espero haberte ayudado.

**Comment by franciscomirasg** [comment]
Gracias por la respuesta
